### PR TITLE
Adds indexTranslationUnit(indexAction:indexerCallbacks:options:)

### PR DESCRIPTION
**Pull request by achrafmam2**

### CODE REVIEW

File: Tests/ClangTests/ClangTests.swift [hunk 1]
```diff
@@ -133,6 +133,30 @@ class ClangTests: XCTestCase {
...
   }
 
+
+  func testIndexAction() {
+    do {
+      let filename = "input_tests/index-action.c"
+      let unit = try TranslationUnit(filename: filename)
+
+      var indexerCallbacks = IndexerCallbacks()
+      indexerCallbacks.indexDeclaration = { _, declPtr in
```
[the remark this answers is not in the document]
Comment (achrafmam2) [reply Tests/ClangTests/ClangTests.swift]: I have tried to do so. The issue is that `IndexerCallbacks` accepts C function pointers which do not capture context. I did not find a mean for unwrapping Swift types to their native types in the passed function.

I am very interested if you have another alternative.

**Comment by harlanhaskins** [comment]
The way we’ve always done this is with a class to wrap each closure, passing that class as the context parameter using `Unmanaged`, then inside the call back (which you can pass a closure to, as long as it doesn’t capture) pulling the Unmanaged reference out, translating the parameters, and calling the closure.

It’s a lot of work, but it’s way worth it in the end for a nice API.

**Comment by achrafmam2** [comment]
@harlanhaskins Done.

File: Sources/Clang/TranslationUnit.swift [hunk 2]
```diff
@@ -411,6 +411,26 @@ public class TranslationUnit {
...
+  /// IndexerCallbacks.
+  /// The order of callback invocations is not guaranteed to be the same as when
+  /// indexing a source file. The high level order will be:
+  /// - Preprocessor callbacks invocations
+  /// - Declaration/reference callbacks invocations
+  /// - Diagnostic callback invocations
+  public func indexTranslationUnit(indexAction: IndexAction,
+                                   indexerCallbacks: IndexerCallbacks,
+                                   options: IndexOptFlags) {
+    // TODO: throw error when clang_indexTranslationUnit returns a non zero value.
```
Comment (harlanhaskins) [inline Sources/Clang/TranslationUnit.swift]: Can you do this? Once this is done it’s good to merge.

Comment (achrafmam2) [reply Sources/Clang/TranslationUnit.swift]: Done.

**Pull request merged**